ocaml/typing/jkind.ml

```diff
@@ -261,6 +274,238 @@ end
 
 type sort = Sort.t
 
+(* A *layout* of a type describes the way values of that type are stored at
+   runtime, including details like width, register convention, calling
+   convention, etc. A layout may be *representable* or *unrepresentable*.  The
+   middle/back ends are unable to cope with values of types with an
+   unrepresentable layout. The only unrepresentable layout is `any`, which is
+   the top of the layout lattice. *)
+module Layout = struct
+  module Const = struct
+    type t =
+      | Sort of Sort.const
+      | Any
+
+    let equal c1 c2 =
+      match c1, c2 with
+      | Sort s1, Sort s2 -> Sort.equal_const s1 s2
+      | Any, Any -> true
+      | (Any | Sort _), _ -> false
+
+    let sub c1 c2 : Sub_result.t =
+      match c1, c2 with
+      | _ when equal c1 c2 -> Equal
+      | _, Any -> Sub
+      | Any, Sort _ | Sort _, Sort _ -> Not_sub
+  end
+
+  type t =
+    | Sort of Sort.t
+    | Any
+
+  let max = Any
+
+  let equate_or_equal ~allow_mutation t1 t2 =
+    match t1, t2 with
+    | Sort s1, Sort s2 -> (
+      match Sort.equate_tracking_mutation s1 s2 with
+      | (Equal_mutated_first | Equal_mutated_second) when not allow_mutation ->
+        Misc.fatal_errorf "Jkind.equal: Performed unexpected mutation"
+      | Unequal -> false
+      | Equal_no_mutation | Equal_mutated_first | Equal_mutated_second -> true)
+    | Any, Any -> true
+    | (Any | Sort _), _ -> false
+
+  let sub t1 t2 : Sub_result.t =
+    match t1, t2 with
+    | Any, Any -> Equal
+    | _, Any -> Sub
+    | Any, _ -> Not_sub
+    | Sort s1, Sort s2 -> if Sort.equate s1 s2 then Equal else Not_sub
+
+  let intersection t1 t2 =
+    match t1, t2 with
+    | _, Any -> Some t1
+    | Any, _ -> Some t2
+    | Sort s1, Sort s2 -> if Sort.equate s1 s2 then Some t1 else None
+
+  let of_new_sort_var () =
+    let sort = Sort.new_var () in
+    Sort sort, sort
+
+  let of_sort_for_error sort = Sort sort
+
+  let value = Sort Sort.value
+
+  let void = Sort Sort.void
+
+  let float64 = Sort Sort.float64
+
+  module Debug_printers = struct
+    open Format
+
+    let t ppf = function
+      | Any -> fprintf ppf "Any"
+      | Sort s -> fprintf ppf "Sort %a" Sort.Debug_printers.t s
+  end
+end
+
+(* Whether or not a type is external to the garbage collector *)
+module Externality = struct
```

Comment:
I've added that CR; you're right that it should be with the other modes. But between the mode solver work and the fact that I wanted to keep this PR self-contained, it was easier to put it here for now.

A mode in a modal kind is an upper bound. That is, suppose expression `e` has a type whose kind includes the mode `m`. Then if `e` is at a mode greater than `m` along `m`'s axis, we can just say `e` is at mode `m`. Since the current `Externality.t` type is used only for mode crossing -- and not for the mode of an expression -- constructors like the ones you're describing make sense. But the goal will be to use the same structure for existing modes, in which case the current constructor names will line up nicely.

Does that help?